File: python/Stack/22. Generate Parentheses.py
'''
BackTracking
if openNum == closedNum == n: add stack to res
if openNum < n: add '(' to stack
if openNum > closedNum: add ')' to stack
'''

class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        stack = []

        def backTrack(openNum, closedNum):
            if openNum == closedNum == n:
                res.append("".join(stack))
                return

            if openNum < n:
                stack.append('(')
                backTrack(openNum + 1, closedNum)
                stack.pop()

            if openNum > closedNum:
                stack.append(')')
                backTrack(openNum, closedNum + 1)
                stack.pop()

        backTrack(0, 0)
        return res



# if n = 3:
# 1:
# ( ( ( ) ) )
# 2:
# ( ( ) ( ) )
# 3:
# ( ( ) ) ( )
# 4
# ( ( ) ) ( )
# 5
# ( ) ( ( ) )
# 6
# ( ) ( ) ( )

File: javascript/Stack/22. Generate Parentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
    let res = [];
    let stack = "";

    const backTrack = (left, right) => {
        if (left === right && right === n) {
            res.push(stack);
            return;
        }
        if (left < n) {
            stack += '(';
            backTrack(left + 1, right);
            stack = stack.slice(0, -1);
        }
        if (left > right) {
            stack += ')';
            backTrack(left, right + 1);
            stack = stack.slice(0, -1);
        }
    }
    backTrack(0, 0);
    return res;
};